R/epichains.R

```diff
@@ -0,0 +1,283 @@
+#' @importFrom utils head
+#' @return object of class `data.frame`
+#' @author James M. Azam
+#' @export
+head.epichains <- function(x, ...) {
+  utils::head(as.data.frame(x), ...)
+}
+
+#' `tail` method for [`epichains`] class
+#' @param x An [`epichains`] object
+#' @param ... further arguments passed to or from other methods
+#' @importFrom utils tail
+#' @author James M. Azam
+#' @export
+tail.epichains <- function(x, ...) {
+  utils::tail(as.data.frame(x), ...)
+}
+
+#' Aggregate cases in epichains objects according to a grouping variable
+#'
+#' @param x An [`epichains`] object.
+#' @param grouping_var The variable to group and count over. Options include
+#' "time", "generation", and "both".
+#' @param ... Other arguments passed to aggregate.
+#' @importFrom stats aggregate
+#' @return If grouping_var is either "time" or "generation", a data.frame
+#' with cases aggregated over `grouping_var`; If
+#' \code{grouping_var = "both"}, a list of data.frames, the first being for
+#'  cases over time, and the second being for cases over generations.
+#' @export
+#'
+#' @examples
+#' set.seed(123)
+#' chains <- simulate_tree(nchains = 10, serials_sampler = function(x) 3,
+#' offspring_sampler = "pois", lambda = 2, chain_stat_max = 10)
+#' chains
+#'
+#' # Aggregate cases per time
+#' aggregate(chains, grouping_var = "time")
+#'
+#' # Aggregate cases per generation
+#' aggregate(chains, grouping_var = "generation")
+#'
+#' # Aggregate cases per both time and generation
+#' aggregate(chains, grouping_var = "both")
+aggregate.epichains <- function(x,
+                                grouping_var = c("time",
+                                                 "generation",
+                                                 "both"
+                                                 ),
+                                ...) {
+  validate_epichains(x)
+  # Check that the object is of type "chains_tree"
+  if (attributes(x)$chain_type != "chains_tree") {
+    stop("object must be an epichains object with 'chains_tree' attribute.")
+  }
+
+  # Get grouping variable
+  grouping_var <- match.arg(grouping_var)
+
+  out <- if (grouping_var == "time") {
+    # Count the number of cases per generation
+    stats::aggregate(list(cases = x$sim_id),
+      list(time = x$time),
+      FUN = NROW
+    )
+  } else if (grouping_var == "generation") {
+    # Count the number of cases per time
+    stats::aggregate(list(cases = x$sim_id),
+      list(generation = x$generation),
+      FUN = NROW
+    )
+  } else if (grouping_var == "both") {
+    # Count the number of cases per time
+    list(
+      stats::aggregate(list(cases = x$sim_id),
+                       list(time = x$time),
+                       FUN = NROW),
+      # Count the number of cases per generation
+      stats::aggregate(list(cases = x$sim_id),
+                       list(generation = x$generation),
+                       FUN = NROW)
+    )
+  }
+
+  structure(out,
```

Comment:
Does this really need its own S3 class? Do you envision any subsequent operations on this? I think it could make sense to just return a `data.frame`

More generally I also wonder if this function is needed (happy to convinced it is!). It seems the same could be achieved with standard R routines such as e.g.

```r
as.data.frame(x) |>
  dplyr::count(generation)
```

Reply:
This method is meant to work under the hood of the `plot()` method. The S3 class was to be passed on to `plot()`, but I later removed `plot`, so it makes this look unnecessary. I plan to reimplement the plotting method, after which, this will probably make more sense. The S3 class is used to also store the grouping variable for easy (auto)plotting.